docs/proposals/20240529-node-level-traffic-reuse-capability.md

```diff
@@ -0,0 +1,109 @@
+| Separate yurt-manager clients | @zyjhtangtang |   @rambohe-ch        | 2024-05-29    |              |        |
+
+---
+
+
+<!-- TOC -->
+* [Node-level Traffic Reuse Capability](#Node-level Traffic Reuse Capability)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals/Future Work](#non-goals)
+  * [Proposal](#proposal)
+  * [User Stories](#user-stories)
+<!-- TOC -->
+
+## Summary
+
+&nbsp;&nbsp;In an OpenYurt cluster, control components are deployed in the cloud, and edge nodes usually interact with the cloud through the public internet, which can lead to significant consumption of cloud-edge traffic. This problem is more pronounced in large-scale clusters, mainly due to the edge-side components performing full-scale list/watch operations on resources. This not only consumes a large amount of cloud-edge traffic but also places considerable pressure on the apiserver due to the high volume of list operations. This proposal presents a solution to optimize this issue.
+
+
+## Motivation
+&nbsp;&nbsp;As illustrated, within an OpenYurt cluster, components such as kubelet, flannel, kube-proxy, coredns, etc., are deployed on each node. YurtHub on the node proxies the component's requests to the apiserver, handling each component's requests independently, which means for every component's list/watch connection, YurtHub maintains a corresponding list/watch connection. When multiple components list/watch the same resource, it leads to data redundancy.
+<img src = "../img/img-20240529001.png" width="800" />
+
+&nbsp;&nbsp;Taking the system components kubelet, coredns, and kubeproxy in OpenYurt as examples, assuming the current scale of the cluster is: 1000 nodes, 10,000 services (each service 0.5KB), and 10,000 endpointslices (each endpointslice 2KB). The traffic generated by these three components for requesting services and endpointslices would be as follows:
+
+- Service traffic: `1000 nodes * 3 components * 10,000 services * 0.5KB/service = 15GB`
+- Endpointslice traffic: `1000 nodes * 2 components * 10,000 endpointslices * 2KB/endpointslice = 40GB`
+### Goals
+
+&nbsp;&nbsp;The optimization described involves reducing the traffic from each node's components to the apiserver. By consolidating the traffic from kubelet, coredns, and proxy, the new service and endpointslice traffic would be:
+
+- Service traffic reduced to: `1000 nodes * 1 consolidated request * 10,000 services * 0.5KB/service = 5GB`
+- Endpointslice traffic reduced to: `1000 nodes * 1 consolidated request * 10,000 endpointslices * 2KB/endpointslice = 20GB`
+
+### Non-Goals/Future Work
+
+- The optimization only involves the reuse of requests on a single node and does not pertain to traffic optimization at the node pool level;
+- The optimization only pertains to full-scale list/watch requests for resources and does not involve requests with filtering conditions.
```

Comment:
I think it is a good idea that requests with filtering conditions can also be supported if full-scale list/wach request exists.

Reply:
Yes, it can support list/watch requests with filter conditions. But it does not support get request.